File: lib/orchestration/debate.ts
import { Message, Participant, DebateResult, DebateRound, DebateTurn, DebateProgressEvent } from './types';
import { getProvider, generateText } from './providers';

export async function runDebate({ history, participants, rounds = 1, topic, onProgress }: {
    history: Message[];
    participants: Participant[];
    rounds?: number;
    topic?: string;
    onProgress?: (event: DebateProgressEvent) => void;
}): Promise<DebateResult> {

    const currentHistory = [...history];
    const debateRounds: DebateRound[] = [];
    
    onProgress?.({ type: 'start', topic, rounds });

    // We need a default API key/provider for the moderator/summarizer if not explicitly provided
    // We'll use the first participant's credentials as a fallback for the moderator
    const primaryParticipant = participants[0];

    for (let i = 0; i < rounds; i++) {
        onProgress?.({ type: 'round_start', roundNumber: i + 1 });
        const turns: DebateTurn[] = [];

        for (const p of participants) {
            onProgress?.({ type: 'turn_start', participantId: p.id, participantName: p.name, role: p.role });
            const provider = getProvider(p.provider);
            if (!provider) {
                console.warn(`Provider ${p.provider} not found for participant ${p.name}`);
                continue;
            }

            const systemPrompt = `You are ${p.name}, acting as a ${p.role}.
            ${topic ? `Topic: ${topic}` : ''}

            Instructions:
            ${p.systemPrompt}

            Review the conversation history and provide your input.
            - Critically analyze previous points.
            - Be constructive, specific, and concise.
            - If you agree, explain why. If you disagree, offer alternatives.`;

            try {
                // Call provider
                const response = await provider.complete({
                    messages: currentHistory,
                    model: p.model,
                    apiKey: p.apiKey,
                    systemPrompt
                });

                const content = response.content;
                const msg: Message = {
                    role: 'assistant',
                    content: content,
                    name: p.id // This helps some providers/clients track who said what
                };

                // Add to history so next participant sees it
                currentHistory.push({
                    role: 'assistant',
                    content: `[${p.name} (${p.role})]: ${content}` 
                });

                turns.push({
                    participantId: p.id,
                    participantName: p.name,
                    role: p.role,
                    content,
                    timestamp: new Date().toISOString()
                });
                
                onProgress?.({ type: 'turn_complete', participantId: p.id, content });

            } catch (err) {
                console.error(`Error processing turn for ${p.name}:`, err);
                const errorContent = `[Error: ${err instanceof Error ? err.message : 'Unknown error'}]`;
                turns.push({
                    participantId: p.id,
                    participantName: p.name,
                    role: p.role,
                    content: errorContent,
                    timestamp: new Date().toISOString()
                });
                onProgress?.({ type: 'turn_complete', participantId: p.id, content: errorContent });
            }
        }

        debateRounds.push({
            roundNumber: i + 1,
            turns
        });
        
        onProgress?.({ type: 'round_complete', roundNumber: i + 1, turns });
    }

    // --- Consensus / Summary Phase ---
    onProgress?.({ type: 'summary_start' });
    let summary = `Debate completed (${rounds} round${rounds > 1 ? 's' : ''}).`;
    
    // Use the primary participant for summary, but check if we have a valid provider first.
    // If not, try to find ANY valid provider from the participants list.
    const summaryParticipant = participants.find(p => getProvider(p.provider)) || primaryParticipant;

    if (summaryParticipant) {
        try {
            const provider = getProvider(summaryParticipant.provider);
            if (!provider) {
                // Should be caught by the find above, but double check
                console.warn("No valid provider found for summary generation");
            } else {
                const moderatorPrompt = `You are the Moderator and Judge of this technical debate.
                ${topic ? `Topic: ${topic}` : ''}
                
                Review the debate history above.
                1. Summarize the key arguments from each participant.
                2. Identify areas of consensus and disagreement.
                3. Provide a final conclusion or recommendation based on the strongest arguments.
                
                Format: Markdown.`;

                // We pass the full history including the new debate turns
                const synthesis = await provider.complete({
                    messages: [
                        ...currentHistory,
                        { role: 'system', content: moderatorPrompt } // Some providers might prefer system role in messages
                        // OR we can append to the last message if 'system' role isn't supported in history array mix
                    ],
                    model: summaryParticipant.model,
                    apiKey: summaryParticipant.apiKey,
                    // valid system prompt if the provider supports separate system arg
                    systemPrompt: moderatorPrompt 
                });
                
                summary = synthesis.content;
            }

        } catch (e) {
            console.error("Failed to generate debate summary:", e);
            summary += " (Auto-summary generation failed)";
        }
    }
    
    onProgress?.({ type: 'summary_complete', summary });

    const result: DebateResult = {
        topic,
        rounds: debateRounds,
        summary,
        history: currentHistory
    };
    
    onProgress?.({ type: 'complete', result });

    return result;
}

export async function runConference({ history, participants, onProgress }: {
    history: Message[];
    participants: Participant[];
    onProgress?: (event: DebateProgressEvent) => void;
}): Promise<DebateResult> {
    // A conference is essentially a single-round debate/discussion
    // where each participant weighs in on the current state.
    // We can reuse the debate logic with rounds=1.
    return runDebate({
        history,
        participants,
        rounds: 1,
        topic: "Team Conference",
        onProgress
    });
}

File: archive/backend-go/services/debate.go
package services

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jules-autopilot/backend/db"
	"github.com/jules-autopilot/backend/models"
)

type DebateMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
	Name    string `json:"name,omitempty"`
}

type DebateParticipant struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	Role         string `json:"role"`
	SystemPrompt string `json:"systemPrompt"`
	Provider     string `json:"provider"`
	Model        string `json:"model"`
	APIKey       string `json:"apiKey,omitempty"`
}

type DebateTurn struct {
	ParticipantID   string    `json:"participantId"`
	ParticipantName string    `json:"participantName"`
	Role            string    `json:"role"`
	Content         string    `json:"content"`
	Timestamp       string    `json:"timestamp"`
	Usage           *LLMUsage `json:"usage,omitempty"`
}

type DebateRound struct {
	RoundNumber int          `json:"roundNumber"`
	Turns       []DebateTurn `json:"turns"`
}

type DebateResult struct {
	ID             string                 `json:"id,omitempty"`
	Topic          string                 `json:"topic,omitempty"`
	Rounds         []DebateRound          `json:"rounds"`
	Summary        string                 `json:"summary,omitempty"`
	History        []DebateMessage        `json:"history"`
	Metadata       map[string]interface{} `json:"metadata,omitempty"`
	TotalUsage     *LLMUsage              `json:"totalUsage,omitempty"`
	RiskScore      int                    `json:"riskScore,omitempty"`
	ApprovalStatus string                 `json:"approvalStatus,omitempty"`
	DurationMs     int64                  `json:"durationMs,omitempty"`
}

type DebateRequest struct {
	Topic        string                 `json:"topic"`
	History      []DebateMessage        `json:"history"`
	Participants []DebateParticipant    `json:"participants"`
	Metadata     map[string]interface{} `json:"metadata,omitempty"`
	Rounds       int                    `json:"rounds,omitempty"`
}

func debateApprovalStatus(score int) string {
	if score < 20 {
		return "approved"
	}
	if score > 80 {
		return "rejected"
	}
	if score > 50 {
		return "flagged"
	}
	return "pending"
}

func normalizeDebateAPIKey(participant DebateParticipant) string {
	apiKey := strings.TrimSpace(participant.APIKey)
	if apiKey != "" && apiKey != "env" && apiKey != "placeholder" {
		return apiKey
	}
	return getSupervisorAPIKey(normalizeProvider(participant.Provider), nil)
}

func RunDebate(request DebateRequest) (DebateResult, error) {
	startTime := time.Now()
	if request.Rounds <= 0 {
		request.Rounds = 1
	}

	currentHistory := make([]DebateMessage, len(request.History))
	copy(currentHistory, request.History)
	rounds := make([]DebateRound, 0, request.Rounds)
	totalUsage := &LLMUsage{}

	for i := 0; i < request.Rounds; i++ {
		turns := make([]DebateTurn, 0, len(request.Participants))
		for _, participant := range request.Participants {
			apiKey := normalizeDebateAPIKey(participant)
			if apiKey == "" {
				turns = append(turns, DebateTurn{
					ParticipantID:   participant.ID,
					ParticipantName: participant.Name,
					Role:            participant.Role,
					Content:         "[Error: Missing API key]",
					Timestamp:       time.Now().Format(time.RFC3339),
				})
				continue
			}

			messages := make([]LLMMessage, 0, len(currentHistory))
			for _, message := range currentHistory {
				messages = append(messages, LLMMessage{Role: message.Role, Content: message.Content})
			}

			systemPrompt := fmt.Sprintf("You are %s, acting as a %s.\n%s\n\nReview the conversation history and provide constructive, specific, concise input.", participant.Name, participant.Role, participant.SystemPrompt)
			result, err := generateLLMText(normalizeProvider(participant.Provider), apiKey, resolveModel(participant.Provider, participant.Model), systemPrompt, messages)
			content := ""
			var usage *LLMUsage
			if err != nil {
				content = fmt.Sprintf("[Error: %s]", err.Error())
			} else {
				content = result.Content
				usage = result.Usage
				if result.Usage != nil {
					totalUsage.PromptTokens += result.Usage.PromptTokens
					totalUsage.CompletionTokens += result.Usage.CompletionTokens
					totalUsage.TotalTokens += result.Usage.TotalTokens
				}
			}

			turns = append(turns, DebateTurn{
				ParticipantID:   participant.ID,
				ParticipantName: participant.Name,
				Role:            participant.Role,
				Content:         content,
				Timestamp:       time.Now().Format(time.RFC3339),
				Usage:           usage,
			})
			currentHistory = append(currentHistory, DebateMessage{
				Role:    "assistant",
				Content: fmt.Sprintf("[%s (%s)]: %s", participant.Name, participant.Role, content),
			})
		}
		rounds = append(rounds, DebateRound{RoundNumber: i + 1, Turns: turns})
	}

	summary := fmt.Sprintf("Debate completed (%d round%s).", request.Rounds, map[bool]string{true: "s", false: ""}[request.Rounds != 1])
	var summaryParticipant *DebateParticipant
	for i := range request.Participants {
		if normalizeDebateAPIKey(request.Participants[i]) != "" {
			summaryParticipant = &request.Participants[i]
			break
		}
	}

	if summaryParticipant != nil {
		messages := make([]LLMMessage, 0, len(currentHistory))
		for _, message := range currentHistory {
			messages = append(messages, LLMMessage{Role: message.Role, Content: message.Content})
		}
		moderatorPrompt := fmt.Sprintf("You are the Moderator and Judge of this technical debate. Topic: %s\nSummarize key arguments, consensus/disagreement, and give a final recommendation in Markdown.", request.Topic)
		apiKey := normalizeDebateAPIKey(*summaryParticipant)
		if result, err := generateLLMText(normalizeProvider(summaryParticipant.Provider), apiKey, resolveModel(summaryParticipant.Provider, summaryParticipant.Model), moderatorPrompt, messages); err == nil && strings.TrimSpace(result.Content) != "" {
			summary = result.Content
			if result.Usage != nil {
				totalUsage.PromptTokens += result.Usage.PromptTokens
				totalUsage.CompletionTokens += result.Usage.CompletionTokens
				totalUsage.TotalTokens += result.Usage.TotalTokens
			}
		}
	}

	riskScore := defaultPlanRiskScore
	approvalStatus := "pending"
	if summaryParticipant != nil {
		apiKey := normalizeDebateAPIKey(*summaryParticipant)
		riskScore = generateRiskScore(normalizeProvider(summaryParticipant.Provider), apiKey, resolveModel(summaryParticipant.Provider, summaryParticipant.Model), request.Topic, summary, defaultPlanRiskScore)
		approvalStatus = debateApprovalStatus(riskScore)
	}

	roundsJSON, _ := json.Marshal(rounds)
	historyJSON, _ := json.Marshal(currentHistory)
	metadataJSON, _ := json.Marshal(request.Metadata)
	workspaceID := "default"
	debateID := uuid.New().String()
	storedSummary := summary
	metadataString := string(metadataJSON)
	debate := models.Debate{
		ID:               debateID,
		Topic:            request.Topic,
		Summary:          &storedSummary,
		Rounds:           string(roundsJSON),
		History:          string(historyJSON),
		Metadata:         &metadataString,
		PromptTokens:     totalUsage.PromptTokens,
		CompletionTokens: totalUsage.CompletionTokens,
		TotalTokens:      totalUsage.TotalTokens,
		WorkspaceID:      &workspaceID,
		CreatedAt:        time.Now(),
		UpdatedAt:        time.Now(),
	}
	if err := db.DB.Create(&debate).Error; err != nil {
		return DebateResult{}, err
	}

	return DebateResult{
		ID:             debateID,
		Topic:          request.Topic,
		Rounds:         rounds,
		Summary:        summary,
		History:        currentHistory,
		Metadata:       request.Metadata,
		TotalUsage:     totalUsage,
		RiskScore:      riskScore,
		ApprovalStatus: approvalStatus,
		DurationMs:     time.Since(startTime).Milliseconds(),
	}, nil
}

func ParseStoredDebate(debate models.Debate) (DebateResult, error) {
	var rounds []DebateRound
	var history []DebateMessage
	var metadata map[string]interface{}
	if strings.TrimSpace(debate.Rounds) != "" {
		_ = json.Unmarshal([]byte(debate.Rounds), &rounds)
	}
	if strings.TrimSpace(debate.History) != "" {
		_ = json.Unmarshal([]byte(debate.History), &history)
	}
	if debate.Metadata != nil && strings.TrimSpace(*debate.Metadata) != "" {
		_ = json.Unmarshal([]byte(*debate.Metadata), &metadata)
	}
	summary := ""
	if debate.Summary != nil {
		summary = *debate.Summary
	}
	return DebateResult{
		ID:       debate.ID,
		Topic:    debate.Topic,
		Rounds:   rounds,
		Summary:  summary,
		History:  history,
		Metadata: metadata,
		TotalUsage: &LLMUsage{
			PromptTokens:     debate.PromptTokens,
			CompletionTokens: debate.CompletionTokens,
			TotalTokens:      debate.TotalTokens,
		},
	}, nil
}
